Pal/lib/slabmgr.h

```diff
@@ -350,10 +341,12 @@ static inline void * slab_alloc (SLAB_MGR mgr, int size)
     system_lock();
     assert(mgr->addr[level] <= mgr->addr_top[level]);
     if (mgr->addr[level] == mgr->addr_top[level] &&
-        listp_empty(&mgr->free_list[level])) {
-        system_unlock();
-        enlarge_slab_mgr(mgr, level);
-        system_lock();
+          listp_empty(&mgr->free_list[level])) {
+        int ret = enlarge_slab_mgr(mgr, level);
+        if (ret < 0) {
+            system_unlock();
+            return NULL;
+        }
```

Comment:
Side note: This change fixes #181.

Reply:
If possible, I would still prefer not to hold lock when allocating memory. I think the real problem of #181 is releasing the lock after adding the new memory to the slab manager.

Reply:
Seems reasonable. `enlarge_slab_mgr` would need some rework to make it race-safe though (I mean: the part before taking the lock).